libmcount/dynamic.c

```diff
@@ -643,6 +643,74 @@ bool mcount_add_badsym(struct mcount_dynamic_info *mdi, unsigned long callsite,
 	return true;
 }
 
+void mcount_handle_dlopen(struct symtabs *symtabs, struct dl_phdr_info *info,
+			  char *mod_realpath, size_t size)
+{
+	struct uftrace_mmap *map;
+	struct mcount_dynamic_info *mdi;
+	char *libname, *modname;
+	unsigned long vaddr = 0, text_addr = 0;
+	unsigned int text_size = 0, namelen, i;
+	bool text_section_found = false;
+
+	modname = basename(mod_realpath);
+
+	if (!find_module_list(modname))
+		return;
+
+	libname = xstrdup(modname);
+	namelen = ALIGN(strlen(libname) + 1, 4);
+	map = xzalloc(sizeof(*map) + namelen);
+	map->start = info->dlpi_addr;
+	map->len = namelen;
+	mcount_memcpy1(map->libname, libname, namelen);
+	map->libname[namelen] = '\0';
+
+	map->end = 0;
+
+	/* find last section and text section address */
+	for (i = 0; i < info->dlpi_phnum; i++) {
+		if (info->dlpi_phdr[i].p_type != PT_LOAD)
+			continue;
+
+		vaddr = ALIGN(info->dlpi_phdr[i].p_vaddr, 0x1000);
+		if (map->end < vaddr)
+			map->end = vaddr;
+
+		if ((info->dlpi_phdr[i].p_flags & PF_X) &&
+		     !text_section_found) {
+			/* find address and size of code segment */
+			text_addr = info->dlpi_phdr[i].p_vaddr;
+			text_size = info->dlpi_phdr[i].p_memsz;
+			text_section_found = true;
+		}
+	}
+
+	map->end += map->start;
+	text_addr += map->start;
+	map->mod = load_module_symtab(symtabs, mod_realpath);
+	maps_add_map(symtabs->maps, map);
+	mdi = create_mdi(map->start, text_addr, text_size);
+
+	if (map->mod) {
+		mdi->map = map;
+		mcount_arch_find_module(mdi, &map->mod->symtab);
+
+		mdi->next = mdinfo;
+		mdinfo = mdi;
+		mdi = setup_trampoline(map);
+		if (mdi == NULL)
+			return;
+
+	}
+	else {
+		free(mdi);
+		return;
+	}
```

Comment:
Maybe we can file earlier than creating `mdi` in this case?

Reply:
you right, and I tihnk it is not need to check null for mdi.

Reply:
Did you change?  I mean we need to update `mdinfo` only if `mcount_setup_trampoline()` is succeeded.

Reply:
because mdi could not be null, removed null check logic.

Reply:
I didn't talk about the NULL check, I'm saying about `mcount_setup_trampoline()`.  If it fails, we probably don't want to have the `mdi` so better to free it.

Reply:
I now understand what you say. it changed.

Reply:
I don't think you did it correctly. Also you freed the `mdi` twice.  What I want is

```c
   mdi = create_mdi(...);
   if (mcount_setup_trampoline(mdi) < 0) {
      free(mdi);
      return;
   }
   mdi->next = mdi;
   mdinfo = mdi;
```